archive/import.go

```diff
@@ -0,0 +1,126 @@
+package archive
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+
+	"github.com/brimsec/zq/driver"
+	"github.com/brimsec/zq/pkg/bufwriter"
+	"github.com/brimsec/zq/pkg/fs"
+	"github.com/brimsec/zq/pkg/nano"
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zio"
+	"github.com/brimsec/zq/zio/zngio"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zqd/api"
+	"github.com/brimsec/zq/zql"
+	"go.uber.org/zap"
+)
+
+func tsDir(ts nano.Ts) string {
+	return ts.Time().Format("20060102")
+}
+
+type importDriver struct {
+	ark *Archive
+	bw  *bufwriter.Writer
+	zw  zbuf.Writer
+	n   int64
+
+	span    nano.Span
+	relpath string
+	spans   []SpanInfo
+}
+
+func (d *importDriver) writeOne(rec *zng.Record) error {
+	recspan := nano.NewSpanTs(rec.Ts, rec.Ts+1)
+	if d.zw == nil {
+		d.span = recspan
+		dir := filepath.Join(d.ark.Root, tsDir(rec.Ts))
+		if err := os.MkdirAll(dir, 0755); err != nil {
+			return err
+		}
+		d.relpath = filepath.Join(tsDir(rec.Ts), rec.Ts.StringFloat()+".zng")
+		path := filepath.Join(d.ark.Root, d.relpath)
+		//XXX for now just truncate any existing file.
+		// a future PR will do a split/merge.
+		out, err := fs.OpenFile(path, os.O_TRUNC|os.O_WRONLY|os.O_CREATE, 0644)
+		if err != nil {
+			return err
+		}
+		d.bw = bufwriter.New(out)
+		d.zw = zngio.NewWriter(d.bw, zio.WriterFlags{})
+	} else {
+		d.span = d.span.Union(recspan)
+	}
+	if err := d.zw.Write(rec); err != nil {
+		return err
+	}
+	d.n += int64(len(rec.Raw))
+	if d.n >= d.ark.Meta.LogSizeThreshold {
+		if err := d.close(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *importDriver) close() error {
+	if d.bw != nil {
+		if err := d.bw.Close(); err != nil {
+			return err
+		}
+		d.spans = append(d.spans, SpanInfo{
+			Span: d.span,
+			Part: d.relpath,
+		})
+	}
+	d.zw = nil
+	d.n = 0
+	return nil
+}
+
+func (d *importDriver) Write(_ int, batch zbuf.Batch) error {
+	for i := 0; i < batch.Length(); i++ {
+		if err := d.writeOne(batch.Index(i)); err != nil {
+			return err
+		}
+	}
+	batch.Unref()
+	return nil
+}
+
+func (d *importDriver) ChannelEnd(cid int) error {
+	return d.close()
```

Comment:
Although this is slick, this is kind of a misuse of the `ChannelEnd` callback- if a forked flowgraph was added this would result in the scanner getting closed preemptively. I would leave this nil (like Warn and Stats) and call d.close after driver.Run finishes.

Reply:
I added panic checks if a non-zero channel id is used; this driver is specifically for the import action below that either uses "sort ts" or "sort -r ts", so it shouldn't be called for multiple channels. PTAL!